### PR TITLE
rgw: update Beast for streaming reads in asio frontend

### PR DESCRIPTION
**Pull request by cbodley**
tracking an upstream Beast development branch that adds support for streaming reads (see issue https://github.com/vinniefalco/Beast/issues/154)

passes all swift functional tests (which cover chunked encoding), and all s3tests except for 12 that send unreadable header values (some tests expect success, and others expect 403). the beast parser rejects these headers, so we reply with 400 Bad Request. you can run s3tests with the tag `!fails_strict_rfc2616` to skip these

for early review - DNM until the development branch is finalized

### CODE REVIEW

**Comment by oritwas** [comment]
jenkins test this please

**Comment by cbodley** [comment]
i was able to run this through teuthology as part of https://github.com/ceph/ceph-ci/commits/wip-rgw-beast-new-parser-qa, which adds an extra commit `qa/rgw: add beast frontend to rgw suites` (i'm leaving that out of this pr until we're committed to maintaining this frontend)

the results with `--filter frontend/beast.yaml` are available at http://pulpito.ceph.com/cbodley-2017-05-01_15:23:40-rgw-wip-rgw-beast-new-parser-qa---basic-mira/:
three unrelated valgrind failures, and one s3tests failure that somehow ran and failed 7 tests that should have been filtered out by `!fails_strict_rfc2616`

File: src/rgw/rgw_asio_client.cc [hunk 1]
```diff
@@ -12,22 +13,23 @@
...
+                   beast::flat_streambuf& buffer)
+  : socket(socket), parser(parser), buffer(buffer), txbuf(*this)
+{
 }
 
-RGWAsioClientIO::~RGWAsioClientIO() = default;
+ClientIO::~ClientIO() = default;
 
-void RGWAsioClientIO::init_env(CephContext *cct)
+void ClientIO::init_env(CephContext *cct)
```
Comment (rzarzynski) [inline src/rgw/rgw_asio_client.cc]: I know it isn't new but maybe:
```cpp
CephContext* cct
```

File: src/rgw/rgw_asio_client.cc [hunk 2]
```diff
@@ -80,42 +82,58 @@ void RGWAsioClientIO::init_env(CephContext *cct)
...
-  body_iter.copy(bytes, buf);
-  return bytes;
+  auto& message = parser.get();
+  auto& body_remaining = message.body;
+  body_remaining = boost::asio::mutable_buffer{buf, max};
+
+  boost::system::error_code ec;
+
+  dout(30) << this << " read_data for " << max << " with "
+      << buffer.size() << " bytes buffered" << dendl;
```
Comment (rzarzynski) [inline src/rgw/rgw_asio_client.cc]: Indentation.

File: src/rgw/rgw_asio_frontend.cc [hunk 3]
```diff
@@ -71,28 +69,47 @@ void Pauser::wait()
...
+      beast::http::async_write(socket, std::move(response), yield[ec]);
+      // ignore ec
       return;
     }
+
+    // process the request
     RGWRequest req{env.store->get_new_req_id()};
-    RGWAsioClientIO real_client{std::move(socket), std::move(request)};
+
+    rgw::asio::ClientIO real_client{socket, parser, buffer};
```
Comment (rzarzynski) [inline src/rgw/rgw_asio_frontend.cc]: OK, `rgw::asio::ClientIO` stores references now.

File: src/rgw/rgw_asio_frontend.cc [hunk 4]
```diff
@@ -7,13 +7,11 @@
 #include <vector>
 
 #include <boost/asio.hpp>
-#include <boost/optional.hpp>
```
Comment (rzarzynski) [inline src/rgw/rgw_asio_frontend.cc]: Thanks!

File: src/rgw/rgw_asio_client.h [hunk 5]
```diff
@@ -4,36 +4,59 @@
 #define RGW_ASIO_CLIENT_H
 
 #include <boost/asio/ip/tcp.hpp>
-#include <beast/http/body_type.hpp>
-#include <beast/http/concepts.hpp>
-#include <beast/http/message_v1.hpp>
+#include <boost/optional.hpp>
```
Comment (rzarzynski) [inline src/rgw/rgw_asio_client.h]: I'm not sure whether the header is really needed.

**Comment by cbodley** [comment]
thanks @rzarzynski. i pushed an update that removes the extra `optional.hpp` include

**Comment by rzarzynski** [comment]
jenkins test this please

**Comment by cbodley** [comment]
new streaming read interfaces merged to Beast's upstream master. i pulled up https://github.com/ceph/Beast to match, and updated the submodule to point there. still building and passing tests

**Comment by smithfarm** [comment]
This breaks the s390x build :-(

```
[  228s] CMake Error at /usr/share/cmake/Modules/FindBoost.cmake:1657 (message):
[  228s]   Unable to find the requested Boost libraries.
[  228s] 
[  228s]   Boost version: 1.63.0
[  228s] 
[  228s]   Boost include path:
[  228s]   /home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495491843.99463ac8f3/build/boost/include
[  228s] 
[  228s] 
[  228s]   Could not find the following static Boost libraries:
[  228s] 
[  228s]           boost_context
[  228s] 
[  228s]   Some (but not all) of the required Boost libraries were found.  You may
[  228s]   need to install these additional Boost libraries.  Alternatively, set
[  228s]   BOOST_LIBRARYDIR to the directory containing Boost libraries or BOOST_ROOT
[  228s]   to the location of Boost.
[  228s] Call Stack (most recent call first):
[  228s]   CMakeLists.txt:595 (find_package)
```

@cbodley Any ideas?

**Comment by smithfarm** [comment]
@cbodley The s390x build is mandatory for us. . . I guess we'll have to conditionalize all the Beast stuff so the s390x builds don't see it. Either that or enable the context module to build on s390x :-/

**Comment by smithfarm** [comment]
Opened http://tracker.ceph.com/issues/20048 to track.

**Comment by smithfarm** [comment]
Testing a fix [1] for the s390x build issue.

[1] https://github.com/ceph/ceph/pull/15225

**Pull request merged**